test/mobile/nnc/test_context.cpp

```diff
@@ -0,0 +1,156 @@
+
+  c10::List<at::Tensor> input({
+      at::ones({size}, at::kFloat).mul(a)
+  });
+  auto outputs = f.run(c10::impl::toList(input));
+  auto output = ((const c10::IValue&) outputs[0]).toTensor();
+  auto expected_output = at::ones({size}, at::kFloat).mul(a * n);
+  EXPECT_TRUE(output.equal(expected_output));
+}
+
+TEST(Function, Serialization) {
+  Function f;
+  f.set_name("test_function");
+  f.set_nnc_kernel_id("test_kernel");
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  f.set_input_specs({create_test_input_spec({1, 3, 224, 224})});
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  f.set_output_spec({create_test_output_spec({1000})});
+  f.set_parameters({
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({1, 16, 3, 3}, at::kFloat),
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({16, 32, 1, 1}, at::kFloat),
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({32, 1, 3, 3}, at::kFloat)
+  });
+  f.set_memory_plan(create_test_memory_plan({
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      sizeof(float) * 1024,
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      sizeof(float) * 2048,
+  }));
+
+  auto serialized = f.serialize();
+  Function f2(serialized);
+  EXPECT_EQ(f2.name(), "test_function");
+  EXPECT_EQ(f2.nnc_kernel_id(), "test_kernel");
+  EXPECT_EQ(f2.input_specs().size(), 1);
+  EXPECT_EQ(f2.input_specs()[0].sizes_, std::vector<int64_t>({1, 3, 224, 224}));
+  EXPECT_EQ(f2.input_specs()[0].dtype_, at::kFloat);
+
+  EXPECT_EQ(f2.output_specs().size(), 1);
+  EXPECT_EQ(f2.output_specs()[0].sizes_, std::vector<int64_t>({1000}));
+  EXPECT_EQ(f2.output_specs()[0].dtype_, at::kFloat);
+
+  EXPECT_EQ(f2.parameters().size(), 3);
+  EXPECT_EQ(f2.parameters()[0].sizes(), at::IntArrayRef({1, 16, 3, 3}));
+  EXPECT_EQ(f2.parameters()[1].sizes(), at::IntArrayRef({16, 32, 1, 1}));
+  EXPECT_EQ(f2.parameters()[2].sizes(), at::IntArrayRef({32, 1, 3, 3}));
+
+  EXPECT_EQ(f2.memory_plan().buffer_sizes_.size(), 2);
+  EXPECT_EQ(f2.memory_plan().buffer_sizes_[0], sizeof(float) * 1024);
+  EXPECT_EQ(f2.memory_plan().buffer_sizes_[1], sizeof(float) * 2048);
+}
+
+TEST(Function, ValidInput) {
+  const int size = 128;
+  Function f;
+  f.set_nnc_kernel_id("dummy");
+  f.set_input_specs({create_test_input_spec({size})});
+
+  c10::List<at::Tensor> input({
+      at::ones({size}, at::kFloat)
+  });
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
+  EXPECT_NO_THROW(
+      f.run(c10::impl::toList(input)));
+}
+
+TEST(Function, InvalidInput) {
+  const int size = 128;
+  Function f;
+  f.set_nnc_kernel_id("dummy");
+  f.set_input_specs({create_test_input_spec({size})});
+
+  c10::List<at::Tensor> input({
+      at::ones({size * 2}, at::kFloat)
+  });
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
+  EXPECT_THROW(
```

Comment:
Should it throw with specific message?

Reply:
It has specific message on the throwing side ("Invalid input at pos..."), I was too lazy to check it in the unit test.
I could add it.

Reply:
Not too important for now.